### ISSUE TITLE
Support vSphere network protocol profiles

### ISSUE THREAD

**Issue by kinolaev** [rancher/rancher#15817]
Hello!

vSphere network protocol profile allows to configure IP pool (and other network options) and pass IP from that pool to guest OS through OVF environment.

Three things needed to support vSphere network protocol profiles:

- [x] `rancher/ui` should generate vApp options rancher/ui#2308
- [x] `rancher/machine` should accept and pass vApp options to vCenter rancher/machine#13
- [x] `rancher/os` should understand new OVF properties rancher/os#2490

![auto](https://user-images.githubusercontent.com/1247759/46900949-537d7980-ceb3-11e8-913d-e5b8c3fa4c8e.png)
![manual](https://user-images.githubusercontent.com/1247759/46900948-52e4e300-ceb3-11e8-84f7-c07a337125dc.png)

vSphere documentation:

- [vSphere vApp properties](https://code.vmware.com/apis/196/vsphere#/doc/vim.vApp.PropertyInfo.html)
- [Network protocol profile](https://docs.vmware.com/en/VMware-vSphere/6.5/com.vmware.vsphere.networking.doc/GUID-4BEEFDA9-E6FF-4D28-B0F3-B02864D8795B.html)

Related forum threads:

https://forums.rancher.com/t/vsphere-k8s-clusters-and-static-ip-assignment-on-nodes/10495
https://forums.rancher.com/t/host-on-vsphere-static-ip-address/5916

**Comment by kinolaev** [comment]
Hello!

Please, pay attention to this PRs. This is very important feature for all vSphere users because ["each node used should have a static IP configured"](https://rancher.com/docs/rancher/v2.x/en/installation/requirements/).

DHCP is bad option because we need manually cleanup lease when removing node. And vSphere already can assign IP address on node creation and release on node deletion, just like other cloud providers.

**Comment by loganhz** [comment]
Thank you so much for your PRs!

I will find someone to review the PR for rancher/machine

The other PRs will be merged after that.

**Comment by kinolaev** [comment]
No! See screenshot and [vSphere vApp properties](https://code.vmware.com/apis/196/vsphere#/doc/vim.vApp.PropertyInfo.html). We need property `guestinfo.interface.0.ip.0.address = ip:NetworkName` and vsphere will replace `ip:NetworkName` with IP from profile's pool when node power on.

**Comment by loganhz** [comment]
Hi @vincent99,  does it make sense now?

**Comment by vincent99** [comment]
Their docs are not very clear about what parts of the type and value strings are literal and which you're supposed to replace, but does `ip:VM Network` actually work, and is that somehow different than `${autoIp:VM Network}`?

I would probably hide all that behind radios for:

```
(x) Do not use vApp
( ) Use vApp to assign a fixed IP from "(the network selected in scheduling section)"
( ) Provide a custom vApp config
```

So that rather than making the user perfectly fill in a dozen fields to enable the one behavior they are likely to want with this feature, they click one radio button.  And if they want something different or want to see what that button does, they can choose custom and get all the controls you have shown.

**Comment by kinolaev** [comment]
Hello @vincent99!

Behavior of `ip:VM Network` dependes on `ipAllocationPolicy` and `${autoIp:VM Network}` doesn't. If `ipAllocationPolicy` is `Transient` then there is no difference.

I agree about radios, when user have only one network it will works. Please confirm that you want it and I will implement it.

**Comment by vincent99** [comment]
Yes, what I said sounds good to me; basically you add a variable in the controller for which mode you're in and tie that to the radio, and observe that mode variable and clear or set your vapp fields appropriately.  On edit of an existing node template you can just always show it as the custom portion.

This (and related backend changes) would go into 2.2.x so I don't see much reason to do the UI in 2 PRs

**Comment by vincent99** [comment]
Does this stuff require particular minimum version, and does it work with bare ESXi or require vSphere? If there are restrictions they should be noted at the top of the new accordion section.

**Comment by kinolaev** [comment]
It is old feature, minimum version is vSphere 4.0. I will try on standalone ESXI next week.

**Comment by kinolaev** [comment]
I made radios, PR rancher/ui#2308 and screenshots updated.

Standalone ESXI doesn't support network protocol profiles because they are created at the level of a data center. "Note: standalone ESXI does not support vApp options and network protocol profiles" added at the top of the new accordion section.

Can I do anything else?
Maybe we don't need to wait until 2.2.x?

**Comment by vincent99** [comment]
It won't go in an actual release until 2.2, but it can be merged into master after reviewing and the backend dependencies are merged.

**Comment by kinolaev** [comment]
Hello @loganhz and @vincent99,

what is the status of this PRs? Is anybody reviewed rancher/machine#13? Any suggestions to improve PRs?

Sorry for annoying, just want to be sure that we will make this feature ready for 2.2.

**Comment by loganhz** [comment]
Thanks @kinolaev 
we will assign someone to review https://github.com/rancher/machine/pull/13 first.

**Comment by kinolaev** [comment]
Hello!

I created PR https://github.com/docker/machine/pull/4598, maybe they can review it.

**Comment by kinolaev** [comment]
Another week is coming to an end without review of rancher/machine#13.

At the DevOops conference, @alena1108  said that `docker-machine` might be replaced with `terraform`. If this is the reason please tell me, I see how to use vApp Options with `terraform`.

**Comment by kinolaev** [comment]
Hello @loganhz!

Do you found someone to review rancher/machine#13? Two months have passed...(

**Comment by loganhz** [comment]
@kinolaev Yes

**Comment by kinolaev** [comment]
Hello @loganhz, @StrongMonkey, @vincent99 and @niusmallnan!

All 3 PRs are now approved. Need I make anything else or you are ready to merge?

**Comment by cjellick** [comment]
Can the rancher (not rancher-os) components be merged and tested without waiting on rancheros?

**Comment by kinolaev** [comment]
Hello @cjellick!

If you will merge ui and machine without os, you will be able to test that vapp properties correctly passed to os, but you will not be able to test static ip configuration because vcenter can't pass ip and netmask in cidr format and os can't use ip and mask separately.

**Comment by cjellick** [comment]
Due to time and resource constraints mainly around testing, we arent going to be able to get this into v2.2.
We appreciate the PRs and will try to get to it early in the 2.3 cycle

**Comment by kinolaev** [comment]
Hello @cjellick!

This is very sad news, considering that PRs were created almost 4 months ago, they have only ~400 changes, they have no breaking changes and they already tested and used in our production system.

RancherOS part is merged and released in 1.5.0, UI and machine parts are reviewed. UI part is relatively big, but machine is only 100 added lines (approved by @StrongMonkey). Maybe you can merge and release machine part in 2.2 and ui part later? This will be good because you have official way to use custom ui, that maybe used until ui part will be released.

**Comment by cjellick** [comment]
Yes, I think I was too hasty in pushing this out. Pulling it back in. The machine changes are isolated enough that I am pretty confident they wont cause any unintended consequences.
Our crunch is in the testing phase, so if we run out of time, we might need to minimize the testing to ensuring there are no regressions.

**Comment by cjellick** [comment]
@kinolaev this issue is in master. you can run master via rancher/rancher:master (not super stable). Want to test it out?

**Comment by cjellick** [comment]
@vicparker to validate (once we get the other vsphere driver update in https://github.com/rancher/rancher/issues/17207)

**Comment by kinolaev** [comment]
Hello @cjellick! I will try to test 2.2 alpha next week.

**Comment by cjellick** [comment]
Cool! The change currently only exists in our `rancher/rancher:master` image (which is a moving target that gets rebuilt and pushed with every commit to master). Feel free to test against that though.

**Comment by kinolaev** [comment]
Hello @cjellick! Feature works except small thing referenced above. Tested with `helm install rancher-latest/rancher --version 2.2.0-rc2`

**Comment by cjellick** [comment]
Tracked here and a PR should be merged this week to fix. https://github.com/rancher/rancher/issues/17712

**Comment by cjellick** [comment]
What is the small thing referenced above?

**Comment by kinolaev** [comment]
I mean https://github.com/rancher/ui/pull/2700, that already merged.

**Comment by vicparker** [comment]
### 2.2.0-rc4 Testing required functionality : vSphere network protocol profile allows to configure IP pool (and other network options) and pass IP from that pool to guest OS through OVF environment
Steps:
1. In vCenter, create Network Profile defining: DNS, Gateway and IP Pool.  
2. Confirm created Network Profile, defined in vCenter, is associated with target network.  
3. Create template (completing vCenter specifications), select Use vApp to configure networks with network protocol profile and Launch.
![image](https://user-images.githubusercontent.com/31374145/54499560-412fbd80-48e1-11e9-8d34-fa6e8d417e21.png)

Verification Steps
1.Tracking the VI cluster creation in the vCenter Task Console, the creation of each node should indicate Network Profile use - profile use is indicated in the details column of the Task Console.  
2. Confirm cluster node's IP configuration matches Network Profile with command: kubectl get nodes -o wide or view nodes in portal and viewing DNS configuration in /etc/resolv.conf.
3. Removing virtual infrastructure working nodes, results in vCenter reclaiming assigned IP address for reuse in IP pool of defined network protocol profile.

**Comment by boliver-pri** [comment]
If you landed on this thread because of generating the same ip address for every node template vm when using VCenter Templates, and you just want to defer to plain ole dhcp, you can fix it by 0'ing out the Machine-ID. Explanation here:

https://anthonyspiteri.net/quick-fix-deploying-multiple-ubuntu-18-04-vms-from-template-with-dhcp-results-same-in-ip-allocation/

**Comment by mfbrj** [comment]
@kinolaev 
@vincent99 
I'm trying without success to use vApp to avoid DHCP in vSphere 6.5. I set this configuration in vSphere:
![Captura de tela_2020-10-21_01-44-08](https://user-images.githubusercontent.com/835834/96674536-66976f80-133f-11eb-9474-5b48b897ed76.png)

In the node template I set:
![image](https://user-images.githubusercontent.com/835834/96674843-03f2a380-1340-11eb-9047-adae8557843e.png)

Unfurtunally, the node created doesn't get an IP address and the node provision get stucked in
`[INFO] [node-controller-rancher-machine] Waiting for SSH to be available...
` 
I even delete everything in the cloud-init config in the node template, and the node still doesn't get an IP.
Am I missing something?

**Comment by marvinlnnx** [comment]
> #17712

I used exactly this steps, it is not working, even I used the [x] Provide a custom vApp config
and set it like so, still not working, I mean the ovaEnv does not have my custom configs:

<img width="546" alt="image" src="https://github.com/user-attachments/assets/d904d8c4-4804-492b-bcec-7bb17491e4fd">



```
com.vmware.guestinfo        IPv4        fixedAllocated
guestinfo.interface.0.ip.0.address      ip:VMware_MGMT
guestinfo.interface.0.ip.0.netmask      ${netmask:VMware_MGMT}
guestinfo.interface.0.route.0.gateway   ${gateway:VMware_MGMT}
guestinfo.dns.servers                   ${dns:VMware_MGMT}
```

`Rancher 2.9.9`
`ubuntu-noble-24.04-cloudimg`
